test/common/buffer/utility.h

```diff
@@ -10,6 +10,36 @@ namespace Envoy {
 namespace Buffer {
 namespace {
 
+/** Used to specify which OwnedImpl implementation to test. */
+enum class BufferImplementation {
+  Old, // original evbuffer-based version
+  New  // new deque-of-slices version
+};
+
+/**
+ * Base class for tests that are parameterized based on BufferImplementation.
+ */
+class BufferImplementationParamTest : public testing::TestWithParam<BufferImplementation> {
+protected:
+  BufferImplementationParamTest() {
+    OwnedImpl::useOldImpl(GetParam() == BufferImplementation::Old);
```

Comment:
independent of coverage testing, how did this even work with just this one file? Wouldn't this have hit your assert?

Does this suggest you were not successfully testing both impls with this structure? It seems you should've been hitting your assert within that one test binary.

Reply:
Prior to the changes I added yesterday, the test code called a different method, `useOldImplForTest`, which didn't do the assert.

The test code has a call to `verifyImplementation(buffer);` in every test case to ensure that it's using the expected implementation.

Reply:
Now that the code allows switching buffer implementations in non-test code, `useOldImplForTest` is no more, and the unit tests just use the same `useOldImpl` that the production code uses.

Reply:
OK, this is fine. My suggestion was based around a theory that we should have safeguards against switching modes except at the start of a binary and between test-methods, but it's really a belt and suspenders argument and as long as you don't actually call it after startup in production it should be fine.